pkg/cri/opts/spec_linux.go

```diff
@@ -139,18 +141,40 @@ func IsCgroup2UnifiedMode() bool {
 }
 
 // WithCDI updates OCI spec with CDI content
-func WithCDI(annotations map[string]string) oci.SpecOpts {
+func WithCDI(annotations map[string]string, CDIDevices []*runtime.CDIDevice) oci.SpecOpts {
 	return func(ctx context.Context, _ oci.Client, c *containers.Container, s *oci.Spec) error {
-		// TODO: Once CRI is extended with native CDI support this will need to be updated...
-		_, cdiDevices, err := cdi.ParseAnnotations(annotations)
+		// Add devices from CDIDevices CRI field
+		var devices []string
+		var err error
+		for _, device := range CDIDevices {
+			devices = append(devices, device.Name)
+		}
+		log.G(ctx).Infof("Container %v: CDI devices from CRI Config.CDIDevices: %v", c.ID, devices)
+
+		// Add devices from CDI annotations
+		_, devsFromAnnotations, err := cdi.ParseAnnotations(annotations)
 		if err != nil {
 			return fmt.Errorf("failed to parse CDI device annotations: %w", err)
 		}
-		if cdiDevices == nil {
-			return nil
+
+		if devsFromAnnotations != nil {
+			log.G(ctx).Infof("Container %v: CDI devices from annotations: %v", c.ID, devsFromAnnotations)
+			for _, deviceName := range devsFromAnnotations {
+				if ctrdutil.InStringSlice(devices, deviceName) {
+					// TODO: change to Warning when passing CDI devices as annotations is deprecated
+					log.G(ctx).Debugf("Skipping duplicated CDI device %s", deviceName)
+					continue
+				}
+				devices = append(devices, deviceName)
+			}
+			// TODO: change to Warning when passing CDI devices as annotations is deprecated
+			log.G(ctx).Debug("Passing CDI devices as annotations is deprecated, please use CRI CDIDevices instead")
```

Comment:
I suppose a question would be who is this warning for.. is this an issue with kubelet/rancher will we ever be able to ignore these...

whatifaboutisms:  what if NRI hook injected annotation.. what if device injected.. 

but yeah let's do a first check on this debug out.. if only because the info log above is more detailed

Reply:
In my understanding we'll be able to deprecate using annotations as soon as kubelet and possibly other CRI counterparts will stop using it. At that point we'll change this debug message to a warning and after some deprecation period we'll stop supporting annotations.